Docs/details/santa-driver.md

```diff
@@ -0,0 +1,93 @@
+
+##### Kauth
+
+santa-driver utilizes two Kauth scopes `KAUTH_SCOPE_VNODE` and `KAUTH_SCOPE_FILEOP `. It registers itself with the Kauth API by calling `kauth_listen_scope()` for each scope. This function takes three arguments.
+
+* `const char *scope`
+* `kauth_scope_callback_t _callback`_
+* `void *contex`
+
+It returns a `kauth_listener_t` that is stored for later use, in Santa's case to stop listening.
+
+###### KAUTH_SCOPE_VNODE
+
+For example here is how santa-driver starts listening for `KAUTH_SCOPE_VNODE` events.
+
+```c++
+vnode_listener_ = kauth_listen_scope(
+    KAUTH_SCOPE_VNODE, vnode_scope_callback, reinterpret_cast<void *>(this));
+```
+
+The function `vnode_scope_callback` is then called for every vnode event. There are many types of vnode events, they complete list can be viewed in the kauth.h. Santa is only concerned with regular files that are generating  `KAUTH_VNODE_EXECUTE ` [1] and `KAUTH_VNODE_WRITE_DATA` events. All non-regular files and unnecessary vnode events are filtered out.
+
+Here is how santa-driver stops listening for `KAUTH_SCOPE_VNODE` events
+
+```c++
+kauth_unlisten_scope(vnode_listener_);
+```
+
+[1] `KAUTH_VNODE_EXECUTE` events that do not have the `KAUTH_VNODE_ACCESS` advisory bit set.
+
+###### KAUTH_SCOPE_FILEOP
+
+Santa also listens for file operations, this is mainly used for logging [1]. 
+
+* `KAUTH_FILEOP_DELETE`, `KAUTH_FILEOP_RENAME`, `KAUTH_FILEOP_EXCHANGE` and `KAUTH_FILEOP_LINK` are logged
+* `KAUTH_FILEOP_EXEC` is used to log `exec()`s. Since the `KAUTH_VNODE_EXECUTE` is used to allow or deny an `exec()` the process arguments have not been setup yet. Since `KAUTH_FILEOP_EXEC` is triggered after an `exec()` it is used to log the `exec()`.
+
+[1] `KAUTH_FILEOP_CLOSE` is used to invalidate that file's representation in the cache. If a file has changed it needs to be re-evalauted. This is particularly necessary for files that were added to the cache with an action of allow.
+
+##### Driver Interface
+
+santa-driver implements a [IOUserClient](https://developer.apple.com/documentation/kernel/iouserclient?language=objc) subclass and santad interacts with it through IOKit/IOKitLib.h functions.
+
+TODO(bur, rah) Flesh out the details
+
+##### Cache
+
+To aid in performance santa-driver utilizes a caching system to hold the state of all observed `exec()` events.
+
+###### Key
+
+The key is a `uint64_t`. The top 32 bits hold the filesystem ID. The bottom 32 bits hold the file unique ID. Together we call this the vnode_id.
```

Comment:
The top 32 bits hold the filesystem ID, while the bottom 32 bits hold the file unique ID.